---
title: "Class 15: Pathway Analysis from RNA-Seq Results"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

##About our input data.
The data for for hands-on session comes from GEO entry: GSE37704, which is associated with the following publication:

> Trapnell C, Hendrickson DG, Sauvageau M, Goff L et al. "Differential analysis of gene regulation at transcript resolution with RNA-seq". Nat Biotechnol 2013 Jan;31(1):46-53. PMID: 23222703

##Import count and metadata
```{r}
metaFile <- "GSE37704_metadata.csv"
countFile <- "GSE37704_featurecounts.csv"

# Import metadata and take a peak
colData = read.csv(metaFile, row.names=1)
head(colData)

```

```{r}
# Import countdata
countData = read.csv(countFile, row.names=1)
head(countData)
```

Hmm... remember that we need the countData and colData files to match up so we will need to remove that odd first column in countData namely contData$length.

```{r}
# Note we need to remove the odd first $length col
countData <- as.matrix(countData[,-1])
head(countData)
```

Double check that the colnames in countData match the id values in the colData metadata file. 

```{r}
colnames(countData)
```

```{r}
rownames(colData)
```

We can use the all() function to check if all entries in a vector are TRUE
```{r}
all(colnames(countData)==rownames(colData))
#Give answer to the question "Are all of them are TRUE?"
```

```{r}
#Test how the all() function works
all(c(T,T,T))
all(c(T,T,F))
```

## Remove any genes with zero count

We want to remove genes that have 0 count values in all experiments (i.e. rows that have 0 across all cols)

```{r}
#rowSums(countData) == 0
```


```{r}
# Filter count data where you have 0 read count across all samples.
countData = countData[rowSums(countData)!=0, ]
head(countData)
```

#DeSeq analysis

Just like R packages from CRAN, you only need to install Bioconductor packages once (instructions here), then load them every time you start a new R session.
```{r}
library(DESeq2)
```

```{r}
# Setup the object with out data in the way DESeq wants it
# DESeq works on a particular type of object called a DESeqDataSet. The DESeqDataSet is a single object that contains input values, intermediate calculations like how things are normalized, and all results of a differential expression analysis.

dds = DESeqDataSetFromMatrix(countData=countData,
                             colData=colData,
                             design=~condition)
# Run the analysis
dds = DESeq(dds)
```

```{r}
dds
```

Get out results
```{r}
res = results(dds)
res
```

```{r}
plot(res$log2FoldChange, -log(res$padj))
```

Lets add some color
```{r}
mycols <- rep("gray",length(res$padj))
mycols[abs(res$log2FoldChange)>2] <- "blue"
mycols[res$padj > 0.005] <- "gray"
plot(res$log2FoldChange, -log(res$padj), col=mycols)
```

## Add gene symbols and entrez ids (for translating language/identifiers)
```{r eval=FALSE}
BiocManager::install("AnnotationDbi")
BiocManager::install("org.Hs.eg.db")

```

```{r}
library("AnnotationDbi")
library("org.Hs.eg.db")

# We can translate between all the following database ID systems 
columns(org.Hs.eg.db) #all the different identifiers that we can map with 
```

```{r}
# Translating our ENSEMBL gene IDs 
res$symbol = mapIds(org.Hs.eg.db,
                    keys=row.names(countData), # where are my IDs
                    keytype="ENSEMBL",         # what format are my IDs
                    column="SYMBOL",           # the new format I want
                    multiVals="first")

```

```{r}
res$entrez = mapIds(org.Hs.eg.db,
                    keys=row.names(countData),
                    keytype="ENSEMBL",
                    column="ENTREZID",
                    multiVals="first")
res$entrez
```

```{r}
res$name =   mapIds(org.Hs.eg.db,
                    keys=row.names(res),
                    keytype="ENSEMBL",
                    column="GENENAME",
                    multiVals="first")
res
```

## Pathway analysis
Here we are going to use the gage package for pathway analysis. Once we have a list of enriched pathways, we're going to use the pathview package to draw pathway diagrams, shading the molecules in the pathway by their degree of up/down-regulation.

```{r}
library(pathview)
library(gage)
library(gageData)
```


```{r}
data(kegg.sets.hs)
data(sigmet.idx.hs)

# Focus on signaling and metabolic pathways only
kegg.sets.hs = kegg.sets.hs[sigmet.idx.hs]

# Examine the first 3 pathways
head(kegg.sets.hs, 3)
```

```{r}
foldchanges = res$log2FoldChange
names(foldchanges) = res$entrez
head(foldchanges)
```

```{r}
# Get the results
keggres = gage(foldchanges, gsets=kegg.sets.hs)
# Now lets look at the object returned from gage().
attributes(keggres)
```


```{r}
#Look at the first few down (less) pathways
head(keggres$less)
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```

```{r}
pathview(gene.data=foldchanges, pathway.id="hsa04110")
```





